### PR TITLE
Solution proposition

### PR DESCRIPTION
**Pull request by eltherion**
I had no write access to the task repo so I forked it and I'll be doing PR's from my repo to original one. I published first PR with the project configuration, API skeleton, some endpoints and (partially) working features.

This is by no means finished. Over the next days and through the progress making I'll sequentially chain next PR's to each other so that if one is being merged, I'll repin next one to the main branch.

### CODE REVIEW

**Comment by pribor** [comment]
Hey @eltherion, could I ask you to implement it in your forked repository and create one PR with a complete solution?

**Comment by eltherion** [comment]
> Hey @eltherion, could I ask you to implement it in your forked repository and create one PR with a complete solution?

@pribor Of course, no problem.

**Comment by eltherion** [comment]
Hello!

Here is the completed PR and solution. Usage info:

https://github.com/GlobalWebIndex/challenge-scala/blob/304aa77af0f02863fefb7382d568a3e5b7cf457c/USAGE.MD

**Comment by pribor** [comment]
Hello, thank you for finishing the task!  At first, I wonder why you decided to mix libraries that don't interop seamlessly together like `akka` vs `cats-eff, fs2`? On the one hand, it is a good demonstration of how to use different stacks together.  On the other hand,  it adds complexity and boilerplate code.

File: src/main/scala/pl/datart/csvtojson/service/TaskRunner.scala [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+class TaskRunnerImpl[F[_]](taskService: TaskService[F], semaphore: Semaphore[F])(implicit
+    async: Async[F]
+) extends TaskRunner[F] {
+  override def run(taskId: TaskId, uri: Uri): F[Unit] = {
+    getAndSave(taskId, uri)
+  }
+
+  private def getAndSave(taskId: TaskId, uri: Uri): F[Unit] = {
+    for {
+      _          <- semaphore.acquire
```
Comment (pribor) [inline src/main/scala/pl/datart/csvtojson/service/TaskRunner.scala]: I think `acquire/release` should have defined boundaries.  It is hard to follow the code where `release` is called from multiple places, and what is more important, it won't be released if e.g.  `taskService.getTask` fails. I would propose to use e.g. `bracket` here or at least `attempt` (won't work with cancellable effects)...

File: src/main/scala/pl/datart/csvtojson/service/Metrics.scala [hunk 2]
```diff
@@ -0,0 +1,5 @@
+package pl.datart.csvtojson.service
+
+import nl.grons.metrics4.scala.DefaultInstrumented
+
+object Metrics extends DefaultInstrumented
```
Comment (pribor) [inline src/main/scala/pl/datart/csvtojson/service/Metrics.scala:5]: Isn't it a global mutable state? What's the point of dealing with `tasks: Ref[F, Map[TaskId, Task]` when we could define it as:

```
object TaskStorage {
   val tasks = TrieMap[TaskId, Task]
}
```   
?

File: src/main/scala/pl/datart/csvtojson/service/TaskService.scala [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+              case Some(t) => t
+            }
+            .takeWhile(!_.isInTerminal, inclusive = true)
+            .map(currentStats)
+            .map(TextMessage(_))
+
+          val sink = Flow[Message]
+            .mapConcat {
+              case binaryMessage: BinaryMessage =>
+                binaryMessage.dataStream.runWith(Sink.ignore)(materializer)
```
Comment (pribor) [inline src/main/scala/pl/datart/csvtojson/service/TaskService.scala]: I think that service shouldn't care about WS, it should only return a `Source`..

**Comment by eltherion** [comment]
> Hello, thank you for finishing the task! At first, I wonder why you decided to mix libraries that don't interop seamlessly together like `akka` vs `cats-eff, fs2`? On the one hand, it is a good demonstration of how to use different stacks together. On the other hand, it adds complexity and boilerplate code.

I did it for few reasons:
 - We talked that the team uses Akka (for example akka-http) during our calls, so I presumed it's kind of expected. :)
 - I used to use akka-http for a longer time than I use http4s (in current company)
 - Having API on akka-http I wanted to keep the rest of the application `F[_]` agnostic and `Future`-free ...
 - ... and in the same time show I can marry both Lightbend/Typesafe and Cats/typelevel.

File: src/main/scala/pl/datart/csvtojson/service/TaskRunner.scala (hunk 1 above)
Comment (eltherion) [reply src/main/scala/pl/datart/csvtojson/service/TaskRunner.scala]: Indeed, it could be rebalanced, which I did.

File: src/main/scala/pl/datart/csvtojson/service/Metrics.scala (hunk 2 above)
Comment (eltherion) [reply src/main/scala/pl/datart/csvtojson/service/Metrics.scala:5]: Well, when using either `DefaultInstrumented` or custom registries its standard usage looks (more or less) like that. I'm referring to:
https://github.com/erikvanoosten/metrics-scala/blob/master/docs/Manual.md#custom-registries

So it introduces state. The line count could be encapsulated in `Task` class and accessed over `tasks` `Ref`, but having it incremented for each line I found it more like _metrics-like_ use case.

File: src/main/scala/pl/datart/csvtojson/service/TaskService.scala (hunk 3 above)
Comment (eltherion) [reply src/main/scala/pl/datart/csvtojson/service/TaskService.scala]: I refactored it. You are right about it.
